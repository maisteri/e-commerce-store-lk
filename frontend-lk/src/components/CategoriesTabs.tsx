import * as React from 'react'
import { Box, Tabs, Tab } from '@mui/material'
import { useAppDispatch, useAppSelector } from '../hooks'
import { getProductsByCategory } from '../reducers/siteGeneralReducer'
import { useNavigate } from 'react-router-dom'

interface CategoriesTabsProps {
  categories: Array<string>
  orientation: 'horizontal' | 'vertical'
  centered: boolean
}

const CategoriesTabs = (props: CategoriesTabsProps) => {
  const [value, setValue] = React.useState(props.categories.length)
  const dispatch = useAppDispatch()
  const categories = useAppSelector((state) => state.general.categories)
  const navigate = useNavigate()

  const handleChange = (event: React.SyntheticEvent, newValue: number) => {
    event.preventDefault()
    setValue(newValue)
    dispatch(getProductsByCategory(categories[newValue]))
    navigate('/')
  }

  return (
    <Box
      sx={{
        width: '100%',
        bgcolor: 'background.paper',
        display:
          props.orientation === 'horizontal'
            ? { xs: 'none', md: 'block' }
            : { xs: 'block' },
      }}
    >
      <Tabs
        value={value}
        onChange={handleChange}
        orientation={props.orientation}
        centered={props.centered}
      >
        {props.categories.map((category) => (
          <Tab label={category} key={category} />
        ))}
      </Tabs>
    </Box>
  )
}

export default CategoriesTabs
