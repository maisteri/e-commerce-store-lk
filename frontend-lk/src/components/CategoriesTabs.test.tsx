// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck

import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import CategoriesTabs from '../components/CategoriesTabs'

const categories = ['cat1', 'cat2', 'cat3']

jest.mock('../reducers/siteGeneralReducer', () => ({
  getProductsByCategory: jest.fn(),
}))

const navigate = jest.fn()
jest.mock('react-router-dom', () => ({
  useNavigate: () => navigate,
}))

const dispatch = jest.fn()
jest.mock('../hooks', () => ({
  useAppDispatch: () => dispatch,
  useAppSelector: () => categories,
}))

describe('<CategoriesTabs />', () => {
  it('renders correctly', () => {
    const { container } = render(
      <CategoriesTabs orientation='horizontal' centered={false} />
    )
    expect(container).toMatchSnapshot()
  })

  it('triggers dispatch and navigation on tab change. Renders updated page after button push.', async () => {
    render(<CategoriesTabs orientation='horizontal' centered={false} />)

    const user = userEvent.setup()
    const button = screen.getByText('cat2')
    await user.click(button)

    expect(navigate.mock.calls).toHaveLength(1)
    expect(dispatch.mock.calls).toHaveLength(1)
    //expect(getProductsByCategory.mock.calls).toHaveLength(1)

    const { container } = render(
      <CategoriesTabs orientation='horizontal' centered={false} />
    )
    expect(container).toMatchSnapshot()
  })
})
