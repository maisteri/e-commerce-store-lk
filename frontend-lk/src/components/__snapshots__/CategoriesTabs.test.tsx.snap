// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CategoriesTabs /> renders correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-1fsq0du"
  >
    <div
      class="MuiTabs-root css-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            cat1
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            cat2
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            cat3
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`<CategoriesTabs /> triggers dispatch and navigation on tab change. Renders updated page after button push. 1`] = `
<div>
  <div
    class="MuiBox-root css-1fsq0du"
  >
    <div
      class="MuiTabs-root css-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
        style="overflow: hidden; margin-bottom: 0px;"
      >
        <div
          class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="0"
            type="button"
          >
            cat1
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            cat2
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-1h9z7r5-MuiButtonBase-root-MuiTab-root"
            role="tab"
            tabindex="-1"
            type="button"
          >
            cat3
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="MuiTabs-indicator css-1aquho2-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;
